File: GetterSetterObjectiveC/GetterSetterObjectiveC/ViewController.m
//
//  ViewController.m
//  GetterSetterObjectiveC
//
//  Created by DMITRY SINYOV on 31.07.17.
//  Copyright © 2017 DMITRY SINYOV. All rights reserved.
//

#import "ViewController.h"
#import "TestableGetSet.h"
@interface ViewController ()

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)beginTestButtonTouched:(id)sender
{
    TestableGetSet *getSet = [[TestableGetSet alloc] init];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self firstWriter:getSet];
    });
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self secondWriter:getSet];
    });
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self reader:getSet];
    });
    
    
//    NSThread *threadWriter1 = [[NSThread alloc] initWithTarget:self selector:@selector(firstWriter:) object:getSet];
//    NSThread *threadWriter2 = [[NSThread alloc] initWithTarget:self selector:@selector(secondWriter:) object:getSet];
//    NSThread *threadReader = [[NSThread alloc] initWithTarget:self selector:@selector(reader:) object:getSet];
//    
//    [threadWriter1 start];
//    [threadWriter2 start];
//    [threadReader start];
//    
//    [threadWriter1 release];
//    [threadWriter2 release];
//    [threadReader release];
    
    [getSet release];
}

-(void)firstWriter:(TestableGetSet*)testObject {
    for (long i = 0; i < 1000; i++) {
        testObject.testField = [NSString stringWithFormat:@"first %ld", i];
    }
}
-(void)secondWriter:(TestableGetSet*)testObject {
    for (long i = 0; i < 1000; i++) {
        testObject.testField = [NSString stringWithFormat:@"second %ld", i];
    }
}
-(void)reader:(TestableGetSet*)testObject {
    for (long i = 0; i < 100; i++) {
        NSLog(@"current %@", testObject.testField);
    }
}


@end

File: GetterSetterSwift/GetterSetterSwift/ViewController.swift
//
//  ViewController.swift
//  GetterSetterSwift
//
//  Created by DMITRY SINYOV on 28.07.17.
//  Copyright © 2017 DMITRY SINYOV. All rights reserved.
//

import UIKit

class ViewController: UIViewController {

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }

    @IBAction func startButtonTaped(_ sender: Any) {
        
        let testObject = TestableGetSet()
        
//        DispatchQueue.global(qos: DispatchQoS.QoSClass.utility).async {
//            self.firstWriter(testObject: testObject)
//        }
//        DispatchQueue.global(qos: DispatchQoS.QoSClass.utility).async {
//            self.secondWriter(testObject: testObject)
//        }
//        DispatchQueue.global(qos: DispatchQoS.QoSClass.utility).async {
//            self.reader(testObject: testObject)
//        }
        
        let threadWriter1 = Thread.init(target: self, selector: #selector(firstWriter(testObject:)), object: testObject)
        let threadWriter2 = Thread.init(target: self, selector: #selector(secondWriter(testObject:)), object: testObject)
        let threadReader = Thread.init(target: self, selector: #selector(reader(testObject:)), object: testObject)
        
        threadWriter1.start()
        threadWriter2.start()
        threadReader.start()
    }
    
    func firstWriter(testObject: TestableGetSet) {
        for i in 0..<1000 {
            testObject.testField = "first \(i)"
        }
    }
    func secondWriter(testObject: TestableGetSet) {
        for i in 0..<1000 {
            testObject.testField = "second \(i)"
        }
    }
    func reader(testObject: TestableGetSet) {
        for _ in 0..<100 {
            print("current \(testObject.testField)")
        }
    }
}

